tree/ntuple/v7/src/RPageStorageDaos.cxx

```diff
@@ -0,0 +1,462 @@
+   WriteNTupleAnchor();
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleHeader(
+		const void *data, size_t nbytes, size_t lenHeader)
+{
+   fContainer->WriteObject(kOidHeader, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenHeader = lenHeader;
+   fNTupleAnchor.fNBytesHeader = nbytes;
+   fNTupleAnchor.fSeekHeader = 0;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleFooter(
+		const void *data, size_t nbytes, size_t lenFooter)
+{
+   fContainer->WriteObject(kOidFooter, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenFooter = lenFooter;
+   fNTupleAnchor.fNBytesFooter = nbytes;
+   fNTupleAnchor.fSeekFooter = 0;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleAnchor() {
+   fContainer->WriteObject(kOidAnchor, &fNTupleAnchor, sizeof(fNTupleAnchor),
+                           kDistributionKey, kAttributeKey);
+}
+
+ROOT::Experimental::Detail::RPage
+ROOT::Experimental::Detail::RPageSinkDaos::ReservePage(ColumnHandle_t columnHandle, std::size_t nElements)
+{
+   if (nElements == 0)
+      nElements = kDefaultElementsPerPage;
+   auto elementSize = columnHandle.fColumn->GetElement()->GetSize();
+   return fPageAllocator->NewPage(columnHandle.fId, elementSize, nElements);
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::ReleasePage(RPage &page)
+{
+   fPageAllocator->DeletePage(page);
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorDaos::NewPage(
+   ColumnId_t columnId, void *mem, std::size_t elementSize, std::size_t nElements)
+{
+   RPage newPage(columnId, mem, elementSize * nElements, elementSize);
+   newPage.TryGrow(nElements);
+   return newPage;
+}
+
+void ROOT::Experimental::Detail::RPageAllocatorDaos::DeletePage(const RPage& page)
+{
+   if (page.IsNull())
+      return;
+   delete[] reinterpret_cast<unsigned char *>(page.GetBuffer());
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPageSourceDaos::RPageSourceDaos(std::string_view ntupleName, std::string_view path,
+   const RNTupleReadOptions &options)
+   : RPageSource(ntupleName, options)
+   , fMetrics("RPageSourceDaos")
+   , fPageAllocator(std::make_unique<RPageAllocatorDaos>())
+   , fPagePool(std::make_shared<RPagePool>())
+   , fPath(path)
+   , fClusterPool(std::make_unique<RClusterPool>(*this))
+{
+   fCounters = std::unique_ptr<RCounters>(new RCounters{
+      *fMetrics.MakeCounter<RNTupleAtomicCounter*>("nReadV", "", "number of vector read requests"),
+      *fMetrics.MakeCounter<RNTupleAtomicCounter*>("nRead", "", "number of byte ranges read"),
+      *fMetrics.MakeCounter<RNTupleAtomicCounter*>("szReadPayload", "B", "volume read from container"),
+      *fMetrics.MakeCounter<RNTuplePlainCounter*> ("szUnzip", "B", "volume after unzipping"),
```

Comment:
Let's follow the latest changes in the file page source and make this counter atomic.